File: c/LeetCode_5_LongestPalindromicSubstring.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


char* expand(char *s, int low, int high) {
    while (low >= 0 && high < strlen(s) && s[low] == s[high]) {
        --low;
        ++high;
    }

    ++low;
    int length = high - low;
    char *word = (char*) malloc((length + 1) * sizeof(char));
    for (int i = 0; i < length; ++i) {
        word[i] = s[low];
        ++low;
    }
    word[length] = '\0';
    return word;
}


char* longestPalindrome(char* s) {
    int length = strlen(s);
    char *result = "";
    char *word = NULL;
    for (int i = 0; i < length; ++i) {
        word = expand(s, i, i);
        if (strlen(result) < strlen(word)) {
            result = word;
        }
        word = expand(s, i, i + 1);
        if (strlen(result) < strlen(word)) {
            result = word;
        }
    }

    return result;
}


int main() {
    char *words = "babad";
    longestPalindrome(words);
    return 0;
}

File: python/LeetCode_5_LongestPalindromicSubstring.py
def expand(s, low, high):
    while (low >= 0 and high < len(s) and s[low] == s[high]):
        low -= 1
        high += 1
    
    return s[low + 1 : high]


def longestPalindrome(s: str) -> str:
    palindrome = ''
    for i in range(len(s)):
        word = expand(i, i)
        if (len(palindrome) < len(word)):
            palindrome = word
        word = expand(i, i + 1)
        if (len(palindrome) < len(word)):
            palindrome = word 

    return palindrome